stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,269 @@ public func gradient<T, U, V, R>(
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool
+  func _isNotEqual(to other: _AnyDerivativeBox) -> Bool
+
+  // `AdditiveArithmetic` requirements.
+  static var _zero: _AnyDerivativeBox { get }
+  static var _cotangentZero: _AnyDerivativeBox { get }
+  func _adding(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _subtracting(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  // `Differentiable` requirements.
+  var _allDifferentiableVariables: _AnyDerivativeBox { get }
+  func _moved(along direction: _AnyDerivativeBox) -> _AnyDerivativeBox
+  func _tangentVector(from cotangent: _AnyDerivativeBox) -> _AnyDerivativeBox
+
+  /// The underlying base value.
+  var _base: Any { get }
+
+  /// Returns the underlying value unboxed to the given type, if possible.
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+
+  /// Returns true if the underlying value has type `AnyDerivative.OpaqueZero`.
+  func _isOpaqueZero() -> Bool
+}
+
+internal struct _ConcreteDerivativeBox<T> : _AnyDerivativeBox
+  where T : Differentiable, T.TangentVector == T,
+        T.AllDifferentiableVariables == T,
+        // NOTE: The requirement below should be defined on `Differentiable`.
+        // But it causes a crash due to generic signature minimization bug.
+        T.CotangentVector == T.CotangentVector.AllDifferentiableVariables
+{
+  /// The underlying base value.
+  var _baseDerivative: T
+
+  public init(_ base: T) {
+    self._baseDerivative = base
+  }
+
+  /// The underlying base value, type-erased to `Any`.
+  var _base: Any {
+    return _baseDerivative
+  }
+
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+  {
+    return (self as _AnyDerivativeBox as? _ConcreteDerivativeBox<U>)?
+      ._baseDerivative
+  }
+
+  func _isOpaqueZero() -> Bool {
+    return T.self == AnyDerivative._OpaqueZero.self
+  }
+
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool {
+    // 0 == 0 => true
+    if _isOpaqueZero() && other._isOpaqueZero() {
+      return true
+    }
+    if _isOpaqueZero() {
+      return type(of: other)._zero._isEqual(to: other)
+    }
+    if other._isOpaqueZero() {
+      return _ConcreteDerivativeBox<T>(T.zero)._isEqual(to: self)
+    }
+    guard let otherBase = other._unboxed(to: T.self) else {
+      fatalError()
```

Comment:
All of these `fatalError()`s should have an error message.

Reply:
Done. Created a `_derivativeTypeMismatch` helper function.

Unfortunately, `print` doesn't display fully qualified type names:
`Derivative type mismatch: _ConcreteDerivativeBox<AllDifferentiableVariables> and _ConcreteDerivativeBox<TangentVector>`

Reply:
I'm a little confused by the text you shared above. Why is it showing the concrete type instead of the `T` type?